### PR TITLE
Support auto-post of combined datasets from registry to portal

### PR DESCRIPTION
**Pull request by sklump**
ckanext-recombinant:
+ recombinant_tables.json supports "target_dataset" per dataset_type (e.g., "ati" for "ati-summaries" and "ati-none")
+ paster recombinant command target-datasets command to return all target datasets defined
+ paster recombinant command dataset-types lists all dataset types per input target

ckanext-canada:
+ ckanext/canada/common code between ati.py, pd.py refactored and generalized in dataset.py
+ ckanext/canada/ati.py, pd.py know their respective target datasets and pass them to dataset.py, which gets corresponding dataset types on the fly
+ bin/reg2portal.sh and reg2portal.py combine dataset resources and upload to corresponding portal datasets
+ bin/csv2solr.sh on portal will update solr core from uploaded csv files, but is still a work in progress

### CODE REVIEW

File: bin/reg2portal.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# Instantiate remote ckanapi
+portal_site = ckanapi.RemoteCKAN(
+    portal_url,
+    apikey=api_key,
+    user_agent=user_agent)
+
+# Update package: add resources
+try:
+    portal_site.call_action(
+        'package_update',
```
Comment (wardi) [inline bin/reg2portal.py]: you want package_patch here instead. update will clear all the other fields too.

File: solr_schema.xml [hunk 2]
```diff
@@ -133,7 +136,7 @@
 
     <field name="_version_" type="string" indexed="true" stored="true"/>
 
-    <dynamicField name="*_date" type="date" indexed="true" stored="true" multiValued="false"/>
+    <!--<dynamicField name="*_date" type="date" indexed="true" stored="true" multiValued="false"/>-->
```
Comment (wardi) [inline solr_schema.xml]: revert the changes to this file, this is the solr schema for ckan not for ATI/PD search

File: bin/csv2solr.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+# Record ckanext-canada home
+cd "${BIN_HOME}/.."
+CXC_HOME=$(pwd)
+
+# Go to ckanext-recombinant
+cd "${BIN_HOME}/../../ckanext-recombinant"
+CXR_HOME=$(pwd)
+
+# Get target dataset names
+TARGET_DATASETS=$(paster recombinant target-datasets -c "${INI_PATH}")
```
Comment (wardi) [inline bin/csv2solr.sh]: Is there a recombinant PR that this relies on?

File: bin/csv2solr.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    for F in ${DATASET_TYPES}
+    do
+      CSV_FILES="${CSV_FILES} ${TARG_DS}.${F}.csv"
+    done
+    CSV_FILES=${CSV_FILES/ /}
+
+    # rebuild solr core from identified .csv files
+    cd "${CXC_HOME}"
+    paster "${TARG_DS}" rebuild -f ${CSV_FILES} -c "${INI_PATH}" ${TARG_DS}.${DS_TYPE}.csv
+done
```
Comment (wardi) [inline bin/csv2solr.sh]: I think this script is complicated enough to warrant rewriting it as a `paster canada` subcommand instead. That way it's not tied to a particular directory layout and can just import and use the necessary recombinant functions.

Comment (wardi) [inline bin/csv2solr.sh]: on second thought I've made a bunch of changes to commands.py for 2.3. It's ok to leave this as is. Just add some docs to the README about how you run the scripts.

File: ckanext/canada/ati.py [hunk 5]
```diff
@@ -55,56 +66,82 @@ def command(self):
...
+                        csv_file,
+                        e.message))
+                except AssertionError as e:
+                    logging.warning('On {0:s}, encountered: {1:s}'.format(
+                        csv_file,
+                        e.message))
+                except IOError as e:
+                    logging.error('On {0:s}, encountered: {1:s}'.format(
+                        csv_file,
+                        e.strerror))
```
Comment (wardi) [inline ckanext/canada/ati.py]: This swallows the tracebacks and only keeps the error message. I'm ok with letting the traceback rise normally and cancel the whole process. We don't want to ignore errors here (and we certainly want all the information necessary to fix them).  Please remove the except: clauses here.

File: bin/csv2solr.sh (hunk 3 above)
Comment (sklump) [inline bin/csv2solr.sh]: Yes - I had neglected it. Please check https://github.com/open-data/ckanext-recombinant/pull/25

**Comment by sklump** [comment]
Ian,

I'm not certain about the solr-schema changes: the solr_schema.ati_pd.xml works on the portal. as is, but there may be less intrusive changes accomplishing the same thing.

I put myself in fits trying to squash and clean up the commit noise, but as the world can plainly see, it got the best of me. Anyway I am going to stop it here before I make things even worse. I was the subversion administrator for Bell Managed Services -- this is embarrassing, but at least the content is good.

**Comment by wardi** [comment]
recombinant work was merged, now working on this pr over here: #451
